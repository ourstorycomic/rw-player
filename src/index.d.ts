import {
  IController,
  IInstance,
  IMedia,
  IModule,
  IModuleLoader,
  IPlayer,
} from './types';

/**
 * Export all the internal types to module developers.
 */
export * from './types';

type ModuleConstructor<T> = new (instance: IInstance) => T;

/**
 * The current version of indigo-player
 * @type {string}
 */
export const VERSION: string;

/**
 * Set the chunks path, this will tell the player where to get the chunks.
 * Defaults to: https://cdn.jsdelivr.net/npm/indigo-player@<VERSION>/lib/
 * @param {string} chunksPath string
 */
export function setChunksPath(chunksPath: string);

/**
 * Registers an external module within the player eco system.
 * @param {IModuleLoader<IModule>} moduleLoader The module loader
 */
export function addModuleLoader(moduleLoader: IModuleLoader<IModule>);

/**
 * Enable console logs.
 * @param {boolean} enableConsoleLogs true/false
 */
export function setConsoleLogs(enableConsoleLogs: boolean);

/**
 * Class constructor for a Module
 * @prop Module
 * @type {ModuleConstructor<IModule>}
 */
export const Module: ModuleConstructor<IModule>;

/**
 * Class constructor for a Controller
 * @prop Controller
 * @type {ModuleConstructor<IController>}
 */
export const Controller: ModuleConstructor<IController>;

/**
 * Class constructor for Media
 * @prop Media
 * @type {ModuleConstructor<IMedia>}
 */
export const Media: ModuleConstructor<IMedia>;

/**
 * Class constructor for Player
 * @prop Player
 * @type {ModuleConstructor<IPlayer>}
 */
export const Player: ModuleConstructor<IPlayer>;

export as namespace IndigoPlayer;
