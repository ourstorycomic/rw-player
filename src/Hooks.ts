import { IHooks, IModule, NextHook } from '@src/types';
import filter from 'lodash/filter';

interface IHook {
  name: string;
  callback: NextHook;
}

/**
 * @Hookable
 * Decorator to let a class know that methods inside can
 * be hooked.
 */
export function Hookable<T extends new (...args: any[]) => {}>(constructor: T) {
  return class extends constructor {
    public hooks = new Hooks((this as unknown) as IModule);
  };
}

class Hooks implements IHooks {
  private module: IModule;

  private hooks: IHook[] = [];

  private origFunctions: any = {};

  constructor(module: IModule) {
    this.module = module;
  }

  public create(name: string, callback: NextHook) {
    this.hookFunction(name);

    this.hooks.push({
      name,
      callback,
    });
  }

  private hookFunction(name: string) {
    if (typeof this.module[name] !== 'function') {
      throw new Error(
        `The method "${name}" does not exist in ${
          this.module.constructor.name
        }`,
      );
    }

    if (this.origFunctions[name]) {
      return;
    }

    // Store the original function and apply a hook.
    this.origFunctions[name] = this.module[name];
    this.module[name] = this.hookedFunction(name);
  }

  private hookedFunction = (name: string) => (...args: any) => {
    const selectedHooks = filter(this.hooks, { name });
    let index = -1;

    const runOrigFunction = () =>
      this.origFunctions[name].call(this.module, ...args);

    const runNextHook = () => {
      const hook = selectedHooks[(index += 1)];

      // If we have no hook to call anymore, call the original function.
      if (!hook) {
        runOrigFunction();
        return;
      }

      let proceed = false;
      const next = () => {
        proceed = true;
      };

      // We've got a hook to call, call it.
      hook.callback.call(null, next, ...args);

      // Did the hook proceed?
      if (proceed) {
        runNextHook();
      }
    };

    runNextHook();
  };
}
