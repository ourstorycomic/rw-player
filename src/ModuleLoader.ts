import { BaseControllerLoader } from '@src/controller/BaseController/BaseControllerLoader';
import { BenchmarkExtensionLoader } from '@src/extensions/BenchmarkExtension/BenchmarkExtensionLoader';
import { ContextMenuExtensionLoader } from '@src/extensions/ContextMenuExtension/ContextMenuExtensionLoader';
import { FreeWheelExtensionLoader } from '@src/extensions/FreeWheelExtension/FreeWheelExtensionLoader';
import { FullscreenExtensionLoader } from '@src/extensions/FullscreenExtension/FullscreenExtensionLoader';
import { GoogleIMAExtensionLoader } from '@src/extensions/GoogleIMAExtension/GoogleIMAExtensionLoader';
import { KeyboardNavigationExtensionLoader } from '@src/extensions/KeyboardNavigationExtension/KeyboardNavigationExtensionLoader';
import { PipExtensionLoader } from '@src/extensions/PipExtension/PipExtensionLoader';
import { StateExtensionLoader } from '@src/extensions/StateExtension/StateExtensionLoader';
import { SubtitlesExtensionLoader } from '@src/extensions/SubtitlesExtension/SubtitlesExtensionLoader';
import { ThumbnailsExtensionLoader } from '@src/extensions/ThumbnailsExtension/ThumbnailsExtensionLoader';
import { DimensionsExtensionLoader } from '@src/extensions/DimensionsExtension/DimensionsExtensionLoader';
import { BaseMediaLoader } from '@src/media/BaseMedia/BaseMediaLoader';
import { DashMediaLoader } from '@src/media/DashMedia/DashMediaLoader';
import { HlsMediaLoader } from '@src/media/HlsMedia/HlsMediaLoader';
import { HTML5PlayerLoader } from '@src/player/HTML5Player/HTML5PlayerLoader';
import {
  IInstance,
  IModule,
  IModuleLoader,
  ModuleLoaderTypes,
} from '@src/types';
import { UiExtensionLoader } from '@src/ui/UiExtensionLoader';
import find from 'lodash/find';

const modules: Array<IModuleLoader<IModule>> = [
  BaseControllerLoader,

  DashMediaLoader,
  HlsMediaLoader,
  BaseMediaLoader,

  HTML5PlayerLoader,

  PipExtensionLoader,
  UiExtensionLoader,
  StateExtensionLoader,
  BenchmarkExtensionLoader,
  FreeWheelExtensionLoader,
  FullscreenExtensionLoader,
  SubtitlesExtensionLoader,
  GoogleIMAExtensionLoader,
  ThumbnailsExtensionLoader,
  KeyboardNavigationExtensionLoader,
  ContextMenuExtensionLoader,
  DimensionsExtensionLoader,
];

export async function createFirstSupported<T>(
  type: ModuleLoaderTypes,
  instance: IInstance,
  isSupportedArgs?: any,
): Promise<T> {
  const items = modules.filter(item => item.type === type);

  for (const loader of items) {
    if (await loader.isSupported(instance, isSupportedArgs)) {
      return ((await loader.create(instance)) as unknown) as T;
    }
  }

  return null;
}

export async function createAllSupported<T>(
  type: ModuleLoaderTypes,
  instance: IInstance,
  isSupportedArgs?: any,
): Promise<T[]> {
  const items = modules.filter(item => item.type === type);

  const instances: T[] = [];

  for (const loader of items) {
    if (await loader.isSupported(instance, isSupportedArgs)) {
      instances.push(((await loader.create(instance)) as unknown) as T);
    }
  }

  return instances;
}

export function addModuleLoader(mod: IModuleLoader<IModule>) {
  modules.push(mod);
}
