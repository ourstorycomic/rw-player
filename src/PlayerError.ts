import { ErrorCodes, IPlayerError } from '@src/types';
import isString from 'lodash/isString';

export class PlayerError extends Error implements IPlayerError {
  public code: ErrorCodes;

  public underlyingError: any;

  constructor(input: ErrorCodes | string, error?: any) {
    super();

    if (isString(input)) {
      this.message = input;
    } else {
      this.code = input as ErrorCodes;
    }

    this.underlyingError = error;
  }
}
