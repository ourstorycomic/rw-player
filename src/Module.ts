import { Hookable } from '@src/Hooks';
import { EventCallback, IEventData, IInstance, IModule } from '@src/types';

@Hookable
export class Module implements IModule {
  public name = 'Unknown';

  public hooks: any;

  public instance: IInstance;

  constructor(instance: IInstance) {
    this.instance = instance;
  }

  public on(name: string, callback: EventCallback) {
    this.instance.on(name, callback);
  }

  public once(name: string, callback: EventCallback) {
    this.instance.on(name, callback);
  }

  public emit(name: string, eventData?: IEventData) {
    this.instance.emit(name, eventData);
  }
}
