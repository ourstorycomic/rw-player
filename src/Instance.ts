import { createConfig } from '@src/createConfig';
import { createAllSupported, createFirstSupported } from '@src/ModuleLoader';
import { PlayerError } from '@src/PlayerError';
import { selectController, selectExtensions } from '@src/selectModule';
import '@src/styles.scss';
import {
  Config,
  ErrorCodes,
  EventCallback,
  Events,
  Format,
  IController,
  IEnv,
  IErrorEventData,
  IEventData,
  IInstance,
  IMedia,
  IModule,
  IPlayer,
  IPlayerError,
  ITrack,
  ModuleLoaderTypes,
} from '@src/types';
import { getEnv } from '@src/utils/getEnv';
import { log } from '@src/utils/log';
import { storage } from '@src/utils/storage';
import EventEmitter from 'eventemitter3';
import find from 'lodash/find';
import isElement from 'lodash/isElement';
import isString from 'lodash/isString';

declare var __webpack_public_path__: string;

export class Instance implements IInstance {
  /**
   * The initial player config.
   * @type {Config}
   */
  public config: Config;

  /**
   * The root container.
   * @type {HTMLElement}
   */
  public container: HTMLElement;

  /**
   * Container specifically for the video element.
   * @type {HTMLElement}
   */
  public playerContainer: HTMLElement;

  public adsContainer: HTMLElement;

  public uiContainer: HTMLElement;

  public env: IEnv;

  public controller: IController;

  public player: IPlayer;

  public media: IMedia;

  public format: Format;

  public extensions: IModule[] = [];

  public log = log;

  public storage = storage;

  /**
   * Allow the instance to emit and listen to events.
   * @type {EventEmitter}
   */
  private emitter: EventEmitter;

  constructor(element: HTMLElement | string, config: Config) {
    this.config = createConfig(config);

    this.createContainers(element, this.config.aspectRatio);

    this.emitter = new EventEmitter();

    this.init(this.config);
  }

  public on = (name: string, callback: EventCallback) =>
    this.emitter.on(name, callback);

  public once = (name: string, callback: EventCallback) =>
    this.emitter.once(name, callback);

  public removeListener = (name: string, callback: EventCallback) =>
    this.emitter.removeListener(name, callback);

  public emit = (name: string, eventData?: IEventData) =>
    this.emitter.emit(name, eventData);

  public play() {
    this.controller.play();
  }

  public pause() {
    this.controller.pause();
  }

  public seekTo(time: number) {
    this.controller.seekTo(time);
  }

  public setVolume(volume: number) {
    this.controller.setVolume(volume);
  }

  public selectTrack(track: ITrack) {
    this.controller.selectTrack(track);
  }

  public selectAudioLanguage(language: string) {
    this.controller.selectAudioLanguage(language);
  }

  public setPlaybackRate(playbackRate: number) {
    this.controller.setPlaybackRate(playbackRate);
  }

  public setError(error: IPlayerError) {
    if (this.controller) {
      this.controller.unload();
    }
    this.emit(Events.ERROR, {
      error,
    } as IErrorEventData);
  }

  public canAutoplay(): boolean {
    return this.config.autoplay && this.env.canAutoplay;
  }

  public destroy() {
    this.emit(Events.DESTROY);

    this.emitter.removeAllListeners();

    this.controller.unload();

    this.controller = null;
    this.player = null;
    this.format = null;
    this.media = null;

    const div: HTMLElement = this.container;
    while (div.firstChild) {
      div.removeChild(div.firstChild);
    }

    this.container.remove();
  }

  public getStats() {
    return {
      config: this.config,
      controller: [this.controller.name, this.controller],
      media: [this.media.name, this.media],
      player: [this.player.name, this.player],
      extensions: this.extensions.map(extension => [extension.name, extension]),
    };
  }

  public getModule(name: string): IModule {
    const modules = [
      ...this.extensions,
      this.controller,
      this.media,
      this.player,
    ];

    return find(modules, { name });
  }

  private createContainers(element: HTMLElement | string, aspectRatio: number) {
    if (isString(element)) {
      element = document.getElementById(element);
    }

    if (!isElement(element)) {
      throw new PlayerError(
        'The provided element is not a valid selector or DOM element.',
      );
    }

    this.container = document.createElement('div');
    this.container.classList.add('ig-container');
    this.container.style.paddingTop = `${100 / aspectRatio}%`;

    this.playerContainer = document.createElement('div');
    this.playerContainer.classList.add('ig-player');
    this.container.appendChild(this.playerContainer);

    this.adsContainer = document.createElement('div');
    this.adsContainer.classList.add('ig-ads');
    this.container.appendChild(this.adsContainer);

    this.uiContainer = document.createElement('div');
    this.uiContainer.classList.add('ig-ui');
    this.container.appendChild(this.uiContainer);

    element.appendChild(this.container);
  }

  private async init(config: Config) {
    this.emit(Events.INSTANCE_INITIALIZE);

    const log = this.log('instance.init');

    log(`Will fetch chunks from ${__webpack_public_path__}`);

    this.env = await getEnv(config);

    this.controller = await selectController(this);
    log('Controller selected', { controller: this.controller });

    this.extensions = await selectExtensions(this);
    log('Extensions loaded', { extensions: this.extensions });

    try {
      await this.controller.load();

      log('Controller loaded');
    } catch (error) {
      if (error instanceof PlayerError) {
        this.setError(error);
      } else {
        this.setError(
          new PlayerError(ErrorCodes.CONTROLLER_LOAD_FAILED, error),
        );
      }
      return;
    }

    this.emit(Events.INSTANCE_INITIALIZED);

    // Set initial config values.
    this.setVolume(config.volume);
    if (config.startPosition) {
      this.seekTo(config.startPosition);
    }

    // Now that we know we can autoplay, actually do it.
    if (this.canAutoplay()) {
      this.play();
      log('play() called because of autoplay');
    }

    setTimeout(() => this.emit(Events.READY));
  }
}
