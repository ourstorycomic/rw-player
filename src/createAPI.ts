import { PlayerError } from '@src/PlayerError';
import { EventCallback, IEventData, IInstance, ITrack } from '@src/types';
import { createFunctionFn } from '@src/utils/defineProperty';

/**
 * Defines the public API, this is the return value of init().
 * @param  {IInstance} instance
 * @return {Object}   External API.
 */
export function createAPI(instance: IInstance) {
  const api: any = {};

  const createFunction = createFunctionFn(api);

  [
    // Bind listeners continuously
    [
      'on',
      (name: string, callback: EventCallback) => instance.on(name, callback),
    ],

    // Bind listeners once
    [
      'once',
      (name: string, callback: EventCallback) => instance.once(name, callback),
    ],

    // Remove a listener
    [
      'removeListener',
      (name: string, callback: EventCallback) =>
        instance.removeListener(name, callback),
    ],

    // Emit an event on the listeners
    [
      'emit',
      (name: string, eventData?: IEventData) => instance.emit(name, eventData),
    ],

    // Play
    ['play', () => instance.play()],

    // Pause
    ['pause', () => instance.pause()],

    // Seek to a time
    ['seekTo', (time: number) => instance.seekTo(time)],

    // Set the volume
    ['setVolume', (volume: number) => instance.setVolume(volume)],

    // Select a track
    ['selectTrack', (track: ITrack) => instance.selectTrack(track)],

    // Select an audio language
    [
      'selectAudioLanguage',
      (language: string) => instance.selectAudioLanguage(language),
    ],

    // Select playback rate
    [
      'setPlaybackRate',
      (playbackRate: number) => instance.setPlaybackRate(playbackRate),
    ],

    // Set a fatal error
    ['setError', (error: PlayerError) => instance.setError(error)],

    // Destroy the player
    ['destroy', () => instance.destroy()],

    // Get stats
    ['getStats', () => instance.getStats()],

    // Get a specific module by name
    ['getModule', (name: string) => instance.getModule(name)],
  ].forEach(tuple => createFunction(tuple[0], tuple[1]));

  api._getInstanceForDev = () => instance;

  return api;
}
