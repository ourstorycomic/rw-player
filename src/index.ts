import { Controller } from '@src/controller/Controller';
import { createAPI } from '@src/createAPI';
import { Instance } from '@src/Instance';
import { Media } from '@src/media/Media';
import { Module } from '@src/Module';
import { addModuleLoader } from '@src/ModuleLoader';
import { Player } from '@src/player/Player';
import { Config, ErrorCodes, Events, ModuleLoaderTypes } from '@src/types';
import { setConsoleLogs } from '@src/utils/log';
import { resolveScriptPath } from '@src/utils/webpack';

declare var __webpack_public_path__: string;
declare var VERSION: string;

__webpack_public_path__ = resolveScriptPath('indigo-player.js');

export default {
  VERSION,
  setChunksPath(chunksPath: string) {
    __webpack_public_path__ = chunksPath;
  },
  setConsoleLogs,
  init(element: HTMLElement | string, config: Config) {
    const instance = new Instance(element, config);
    return createAPI(instance);
  },
  addModuleLoader,

  // Export enums
  Events,
  ErrorCodes,
  ModuleLoaderTypes,

  // Export class constructors
  Module,
  Controller,
  Media,
  Player,
};
