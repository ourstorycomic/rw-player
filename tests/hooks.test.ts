import { Hookable } from '../src/Hooks';

@Hookable
class Mock {
  public sayHello() {}
}

let mock;

beforeEach(() => {
  mock = new Mock();
});

test('can access hooks', () => {
  expect(mock.hooks).toBeDefined();
});

test('can access hooks.create', () => {
  expect(mock.hooks.create).toBeInstanceOf(Function);
});

test('can install a hook that exists', () => {
  const callback = jest.fn();

  mock.hooks.create('sayHello', callback);

  expect(mock.hooks.hooks.length).toBe(1);
  expect(mock.hooks.hooks[0]).toMatchSnapshot();
  expect(mock.hooks.hooks[0].callback).toBe(callback);
});

test('can install multiple hooks', () => {
  mock.hooks.create('sayHello', jest.fn());
  mock.hooks.create('sayHello', jest.fn());
  mock.hooks.create('sayHello', jest.fn());
  expect(mock.hooks.hooks.length).toBe(3);
});

test('installing a hook overwrites the original function', () => {
  const orig = mock.sayHello;
  mock.hooks.create('sayHello', jest.fn());
  expect(mock.sayHello).not.toBe(orig);
});

test('fails when installing a hook that does not exist', () => {
  expect(() => {
    mock.hooks.create('screamHello', jest.fn());
  }).toThrowError();
});

test('invokes a hook if the original function is invoked', () => {
  const callback = jest.fn();
  mock.hooks.create('sayHello', callback);
  mock.sayHello();
  expect(callback).toHaveBeenCalled();
});

test('an invoked hook has a next function', () => {
  const callback = jest.fn();
  mock.hooks.create('sayHello', callback);
  mock.sayHello();
  expect(callback).toHaveBeenCalledWith(expect.any(Function));
});

test('an invoked hook includes additional parameters after the next parameter', () => {
  const callback = jest.fn();
  mock.hooks.create('sayHello', callback);
  mock.sayHello(1, 'hello', true);
  expect(callback).toHaveBeenCalledWith(expect.any(Function), 1, 'hello', true);
});

test('calls the orig function after a hook calls next', () => {
  const spy = jest.spyOn(mock, 'sayHello');
  mock.hooks.create('sayHello', next => next());
  mock.sayHello();
  expect(spy).toHaveBeenCalled();
});

test('does not call the orig function after a hook does not call next', () => {
  const spy = jest.spyOn(mock, 'sayHello');
  mock.hooks.create('sayHello', next => {});
  mock.sayHello();
  expect(spy).not.toHaveBeenCalled();
});

test('calls the orig function after multiple hooks call next', () => {
  const spy = jest.spyOn(mock, 'sayHello');
  mock.hooks.create('sayHello', next => next());
  mock.hooks.create('sayHello', next => next());
  mock.hooks.create('sayHello', next => next());
  mock.sayHello();
  expect(spy).toHaveBeenCalled();
});

test('does not call the orig function after one hook does not call next', () => {
  const spy = jest.spyOn(mock, 'sayHello');
  mock.hooks.create('sayHello', next => next());
  mock.hooks.create('sayHello', next => {});
  mock.hooks.create('sayHello', next => next());
  mock.sayHello();
  expect(spy).not.toHaveBeenCalled();
});
