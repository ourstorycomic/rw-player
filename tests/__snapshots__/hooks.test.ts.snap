// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can install a hook that exists 1`] = `
Object {
  "callback": [MockFunction],
  "name": "sayHello",
}
`;
